packages/client/src/components/UserContext.tsx

```diff
@@ -0,0 +1,39 @@
+import { gameIO } from 'lib/socket';
+import { useEffect } from 'react';
+import { useState } from 'react';
+import { FC, createContext, useContext } from 'react';
+import { getPersistedUserName, persistUserName } from 'utils/localStorageUtils';
+
+export const UserContext = createContext<{
+  userName: string | null;
+}>({
+  userName: null,
+});
+
+const persistedUserName = getPersistedUserName();
+export const UserProvider: FC = ({ children }) => {
+  const [userName, setUserName] = useState<string | null>(persistedUserName);
+  if (!userName) {
+    gameIO.emit('username-missing');
+  }
+
+  useEffect(() => {
+    gameIO.on('connect', () => {
+      console.log(`connect ${gameIO.id}`);
+    });
+    const handleUsernameRequested = (usrName: string) => {
```

Comment:
Is it a good practice to have this nested? Just wondering, I didn't know when I did my stuff.

Reply:
Nested means inside the use effect? It does because at least you don't have to declare it as dependency for the hook